src/components/OptionsSelector/BaseOptionsSelector.js

```diff
@@ -61,622 +54,461 @@ const defaultProps = {
-                if (this.relatedTarget && ref === this.relatedTarget) {
-                    this.textInput.focus();
-                    this.relatedTarget = null;
-                }
-                if (this.textInput.isFocused()) {
-                    setSelection(this.textInput, 0, this.state.value.length);
-                }
-            }
-            const selectedOption = this.props.onSelectRow(option);
-            resolve(selectedOption);
+    useEffect(() => {
+        if (_.isEqual(allOptions, prevOptions)) {
+            return;
+        }
 
-            if (!this.props.canSelectMultipleOptions) {
-                return;
-            }
+        const newFocusedIndex = props.selectedOptions.length;
+        const prevFocusedOption = prevOptions[focusedIndex];
+        const indexOfPrevFocusedOptionInCurrentList = _.findIndex(allOptions, (option) => prevFocusedOption && option.keyForList === prevFocusedOption.keyForList);
+        setFocusedIndex(indexOfPrevFocusedOptionInCurrentList || (_.isNumber(props.focusedIndex) ? props.focusedIndex : newFocusedIndex));
+        // we want to run this effect only when the sections change
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [allOptions]);
+
+    useEffect(() => {
+        // If we just toggled an option on a multi-selection page or cleared the search input, scroll to top
+        if (props.selectedOptions.length !== prevSelectedOptions.length || (!!prevValue && !value)) {
+            scrollToIndex(0);
+            return;
+        }
 
-            // Focus the first unselected item from the list (i.e: the best result according to the current search term)
-            this.setState({
-                focusedIndex: this.props.selectedOptions.length,
-            });
-        });
-    }
+        // Otherwise, scroll to the focused index (as long as it's in range)
+        if (allOptions.length <= focusedIndex) {
+            return;
+        }
+        scrollToIndex(focusedIndex);
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [allOptions.length, focusedIndex, props.focusedIndex, props.selectedOptions, value]);
+
+    const updateSearchValue = useCallback(
+        (searchValue) => {
+            setValue(searchValue);
+            setErrorMessage(
+                searchValue.length > props.maxLength
+                    ? translate('common.error.characterLimitExceedCounter', {
+                          length: searchValue.length,
+                          limit: props.maxLength,
+                      })
+                    : '',
+            );
+            props.onChangeText(searchValue);
+        },
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+        [props.onChangeText, props.maxLength, translate],
+    );
+
+    const debouncedUpdateSearchValue = _.debounce(updateSearchValue, CONST.TIMING.SEARCH_OPTION_LIST_DEBOUNCE_TIME);
```

Comment:
this creates new debouncedUpdateSearchValue on every render. We have a `useDebouncedState` hook. Then you could move the errorMessage  (and I guess  props.onChangeText too) to a useMemo hook with dep of `debouncedSearchValue`